common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -550,6 +558,197 @@ public static UTF8String toTitleCase(final UTF8String target, final int collatio
+    put(64256, "Ff");
+    put(64257, "Fi");
+    put(64258, "Fl");
+    put(64259, "Ffi");
+    put(64260, "Ffl");
+    put(64261, "St");
+    put(64262, "St");
+    put(64275, "Մն");
+    put(64276, "Մե");
+    put(64277, "Մի");
+    put(64278, "Վն");
+    put(64279, "Մխ");
+  }};
+
+  /**
+   * Title casing a string according to a new behaviour. Iterates over the string and title cases
+   * the first character in each word, and lowercases every other character. Handles lowercasing
+   * capital Greek letter sigma ('Σ') separately, taking into account if it should be a small final
+   * Greek sigma ('ς') or small non-final Greek sigma ('σ'). Words are separated by ASCII
+   * space(\u0020).
+   *
+   * @param target UTF8String to be title cased
+   * @return title cased target
+   */
+  public static UTF8String toTitleCaseICU(UTF8String target) {
+
+    // In the default UTF8String implementation, `toLowerCase` method implicitly does UTF8String
+    // validation (replacing invalid UTF-8 byte sequences with Unicode replacement character
+    // U+FFFD), but now we have to do the validation manually.
+    target = target.makeValid();
+
+    // Building the title cased target with 'sb'.
+    StringBuilder sb = new StringBuilder();
+
+    // 'newWord' is true if the current character is the beginning of a word, false otherwise.
+    boolean newWord = true;
+    // We are maintaining if the current character is preceded by a cased letter.
+    // This is used when lowercasing capital Greek letter sigma ('Σ'), to figure out if it should be
+    // lowercased into σ or ς.
+    boolean precededByCasedLetter = false;
+
+    // 'offset' is a byte offset in target's byte array pointing to the beginning of the character
+    // that we need to process next.
+    int offset = 0;
+    int len = target.numBytes();
+
+    while(offset < len) {
+      // We will actually call 'codePointFrom()' 2 times for each character in the worst case (once
+      // here, and once in 'followedByCasedLetter'). Example of a string where we call it 2 times
+      // for almost every character is 'ΣΣΣΣΣ' (a string consisting only of Greek capital sigma)
+      // and 'Σ`````' (a string consisting of a Greek capital sigma, followed by case-ignorable
+      // characters).
+      int codepoint = target.codePointFrom(offset);
+      // Appending the correctly cased character onto 'sb'.
+      appendTitleCasedCodepoint(sb, codepoint, newWord, precededByCasedLetter, target, offset);
+      // Updating 'newWord', 'precededByCasedLetter' and 'offset' to be ready for the next character
+      // that we will process.
+      newWord = (codepoint == ASCII_SPACE_CODEPOINT);
+      if(!UCharacter.hasBinaryProperty(codepoint, UProperty.CASE_IGNORABLE)) {
+        precededByCasedLetter = UCharacter.hasBinaryProperty(codepoint, UProperty.CASED);
+      }
+      offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+    }
+    return UTF8String.fromString(sb.toString());
```

Comment:
we may be able to use `UTF8StringBuilder` here, provided that we extend it with an `appendCodePoint` method

Reply:
(of course this would incur some small changes to the private helped methods below)